arangod/Aql/Optimizer/ExpressionMatcher/MatchResult.h

```diff
@@ -0,0 +1,103 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2023 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+////////////////////////////////////////////////////////////////////////////////
+#pragma once
+
+#include <string>
+#include <unordered_map>
+#include <variant>
+#include <optional>
+
+#include <Aql/AstNode.h>
+
+namespace arangodb::aql::expression_matcher {
+struct MatchResult {
+  auto isSuccess() const noexcept -> bool { return _errors.empty(); };
+  auto isError() const noexcept -> bool { return not _errors.empty(); }
+
+  struct ErrorTag {};
+  struct ValueTag {};
+
+  explicit MatchResult() = default;
+  MatchResult(MatchResult&&) = default;
+  MatchResult(MatchResult const&) = delete;
+
+  auto addError(std::string error) -> MatchResult&& {
+    _errors.emplace_back(error);
+    return std::move(*this);
+  }
+
+  auto addErrorIfError(std::string error) -> MatchResult&& {
+    if (isError()) {
+      _errors.emplace_back(error);
+    }
+    return std::move(*this);
+  }
+
+  auto addMatch(std::string name, AstNode const* value) -> MatchResult&& {
+    _matches.emplace(name, value);
+    return std::move(*this);
+  }
+
+  auto combine(MatchResult&& rhs) -> MatchResult&& {
+    _errors.insert(std::end(_errors), std::begin(rhs._errors),
+                   std::end(rhs._errors));
+    _matches.merge(rhs._matches);
+
+    return std::move(*this);
+  }
```

Comment:
Since these return an rvalue ref to this, they should probably have an rvalue ref-qualifier
```suggestion
  auto addError(std::string error) && -> MatchResult&& {
    _errors.emplace_back(error);
    return std::move(*this);
  }

  auto addErrorIfError(std::string error) && -> MatchResult&& {
    if (isError()) {
      _errors.emplace_back(error);
    }
    return std::move(*this);
  }

  auto addMatch(std::string name, AstNode const* value) && -> MatchResult&& {
    _matches.emplace(name, value);
    return std::move(*this);
  }

  auto combine(MatchResult&& rhs) && -> MatchResult&& {
    _errors.insert(std::end(_errors), std::begin(rhs._errors),
                   std::end(rhs._errors));
    _matches.merge(rhs._matches);

    return std::move(*this);
  }
```